File: src/main/java/cptn/util/FileUtil.java
package cptn.util;

import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.PropertyResourceBundle;
import java.util.ResourceBundle;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * 文件操作工具类
 */
public class FileUtil {
	public static final int FILE_BUFF_SIZE = 65536;
	public static final int PROGRESS_ACTION_ID = 900;
	
	private static Logger logger = LoggerFactory.getLogger(DateUtil.class);

	/**
	 * 获取程序运行的当前目录
	 * @return
	 */
	public static String getCurDir() {
		String dir = System.getProperty("user.dir");
		return dir;
	}

	/**
	 * 读取指定资源文件（需指定确定的文件名，包含扩展名，无法自动国际化）
	 * @param filePath
	 * @return
	 */
	public static ResourceBundle getResourceBundle(String filePath) {
		try {
			FileInputStream fis = new FileInputStream(filePath);
			return new PropertyResourceBundle(fis);
		}
		catch (Exception e) {
			logger.error("读取配置文件失败：" + filePath, e);
			return null;
		}
	}

	/**
	 * 创建目录树
	 * @param path
	 * @return
	 */
	public static boolean createFolderTree(String path) {
		boolean result = true;

		try {
			File f = new File(path);

			if (f.exists()) {
				return true;
			}

			f.mkdirs();
		}
		catch (Exception e) {
			logger.warn("", e);
			result = false;
		}

		return result;
	}

	/**
	 * 复制文件
	 * @param is
	 * @param destFile
	 * @param fileSize
	 * @param listener
	 * @return
	 */
	public static int copyFile(InputStream is, String destFile, long fileSize, ActionListener listener) {
		int result = 0;

		try {
			OutputStream out = new FileOutputStream(destFile);
			byte buf[] = new byte[FILE_BUFF_SIZE];
			int len;
			int progress = 0;
			long writedLen = 0;

			while ((len = is.read(buf)) > 0) {
				out.write(buf, 0, len);
				writedLen += len;
				
				if (listener != null && fileSize > 0) {
					progress = (int) (writedLen * 100 / fileSize);
					
					ActionEvent ae = new ActionEvent(FileUtil.class, PROGRESS_ACTION_ID, String.valueOf(progress));
					listener.actionPerformed(ae);
				}
			}

			out.close();
			is.close();
		}
		catch (Exception e) {
			logger.warn("", e);
			result = -1;
		}

		return result;
	}

	/**
	 * 从路径中分离出文件名
	 * @param filePathName
	 * @return
	 */
	public static String getFileName(String filePathName) {
		if (StrUtil.isEmpty2(filePathName)) {
			return "";
		}

		filePathName = filePathName.trim();

		int pos1 = filePathName.lastIndexOf("/");
		int pos2 = filePathName.lastIndexOf("\\");
		int pos = pos1 > pos2 ? pos1 : pos2;

		if (pos >= 0) {
			return filePathName.substring(pos + 1);
		}
		else {
			return filePathName;
		}
	}
	
	/**
	 * 从路径中分离出路径和文件名
	 * @param filePathName
	 * @return
	 */
	public static String[] getFilePathName(String filePathName) {
		String[] data = { "", "" };
		
		if (StrUtil.isEmpty2(filePathName)) {
			return data;
		}

		filePathName = filePathName.trim();

		// 寻找最后一个分隔符
		int pos1 = filePathName.lastIndexOf("/");
		int pos2 = filePathName.lastIndexOf("\\");
		int pos = pos1 > pos2 ? pos1 : pos2;

		// 
		if (pos == (filePathName.length() - 1)) {
			// 以分隔符结尾，一定是路径，不带文件名
			data[0] = filePathName;
			data[1] = "";
		}
		else if (pos >= 0) {
			data[0] = filePathName.substring(0, pos + 1);
			data[1] = filePathName.substring(pos + 1);
		}
		else {
			data[0] = "";
			data[1] = filePathName;
		}
		
		return data;
	}

	/**
	 * 取得文件扩展名
	 * @param pathFileName
	 * @return
	 */
	public static String getFileExt(String pathFileName) {
		String fileName = getFileName(pathFileName);

		if (StrUtil.isEmpty2(fileName)) {
			return "";
		}

		int pos = pathFileName.lastIndexOf(".");

		if (pos >= 0) {
			return pathFileName.substring(pos + 1);
		}
		else {
			return "";
		}
	}

	/**
	 * 删除多个文件
	 * @param files
	 */
//	public static void deleteFiles(String files) {
//
//	}
	
	/**
	 * 在路径后拼接文件名
	 * @param path
	 * @param fileName
	 * @return
	 */
	public static String appendFileName(String path, String fileName) {
		path = StrUtil.safeTrim2(path);
		fileName = StrUtil.safeTrim2(fileName);
		
		boolean b1 = false;
		boolean b2 = false;
		
		if (StrUtil.isEmpty(path)) {
			return fileName;
		}
		
		if (StrUtil.isEmpty(fileName)) {
			return path;
		}
		
		String suffix = path.substring(path.length() - 1);
		if (suffix.equals("/") || suffix.equals("\\")) {
			b1 = true;
		}
		
		String prefix = fileName.substring(0, 1);
		if (prefix.equals("/") || prefix.equals("\\")) {
			b2 = true;
		}
		
		if (b1 && b2) {
			return path + fileName.substring(1);
		}
		else if (b1 || b2) {
			return path + fileName;
		}
		else {
			return path + "/" + fileName;
		}
	}
	
	/**
	 * 为文件名添加扩展名
	 * @param fileName
	 * @param ext
	 * @return
	 */
	public static String appendFileExt(String fileName, String ext) {
		fileName = StrUtil.safeTrim2(fileName);
		ext = StrUtil.safeTrim2(ext);
		
		if (StrUtil.isEmpty2(ext)) {
			return fileName;
		}
		
		return (fileName + "." + ext);
	}
	
	/**
	 * 删除单个文件
	 * @param filePathName
	 * @return
	 */
	public static boolean deleteFile(String filePathName) {
		File file = new File(filePathName);
		
		if (file.exists()) {
			if (!file.isDirectory()) {
				file.delete();
				return true;
			}
		}
		
		return false;
	}
	
	/**
	 * 读取文本文件
	 * @param filePathName
	 * @return
	 * @throws FileNotFoundException
	 * @throws IOException
	 */
	public static List<String> readTextFile(String filePathName) throws FileNotFoundException, IOException {
		List<String> lines = new ArrayList<String>();
		InputStreamReader reader = null;
		BufferedReader br = null;
		InputStream input = null;
		
		try {
			input = new FileInputStream(filePathName);
			reader = new InputStreamReader(input);
	        br = new BufferedReader(reader);  
	        String line;  
	        
	        while ((line = br.readLine()) != null) {  
	            lines.add(line);  
	        }
		}
		finally {
			try {
				if (br != null) {
					br.close();
				}
			} catch (IOException e) {
			}
			
			try {
				if (reader != null) {
					reader.close();
				}
			} catch (IOException e) {
			}

			try {
				if (input != null) {
					input.close();
				}
			} catch (IOException e) {
			}
		}
		
		return lines;
	}
}

File: src/main/java/cptn/util/FileUtil.kt
package cptn.util

import org.slf4j.LoggerFactory
import java.awt.event.ActionEvent
import java.awt.event.ActionListener
import java.io.*
import java.util.*

/**
 * 文件操作工具类
 */
object FileUtil {
    val FILE_BUFF_SIZE = 65536
    val PROGRESS_ACTION_ID = 900

    private val logger = LoggerFactory.getLogger(FileUtil::class.java)

    /**
     * 获取程序运行的当前目录
     * @return
     */
    val curDir: String
        get() = System.getProperty("user.dir")

    /**
     * 读取指定资源文件（需指定确定的文件名，包含扩展名，无法自动国际化）
     * @param filePath
     * @return
     */
    fun getResourceBundle(filePath: String): ResourceBundle? {
        try {
            val fis = FileInputStream(filePath)
            return PropertyResourceBundle(fis)
        } catch (e: Exception) {
            logger.error("读取配置文件失败：$filePath", e)
            return null
        }

    }

    /**
     * 创建目录树
     * @param path
     * @return
     */
    fun createFolderTree(path: String): Boolean {
        var result = true

        try {
            val f = File(path)

            if (f.exists()) {
                return true
            }

            f.mkdirs()
        } catch (e: Exception) {
            logger.warn("", e)
            result = false
        }

        return result
    }

    /**
     * 复制文件
     * @param `ins`
     * @param destFile
     * @param fileSize
     * @param listener
     * @return
     */
    fun copyFile(ins: InputStream, destFile: String, fileSize: Long, listener: ActionListener?): Int {
        var result = 0

        try {
            val out = FileOutputStream(destFile)
            val buf = ByteArray(FILE_BUFF_SIZE)
            var progress = 0
            var writedLen: Long = 0
            var len: Int = ins.read(buf)

            while (len > 0) {
                out.write(buf, 0, len)
                writedLen += len.toLong()

                if (listener != null && fileSize > 0) {
                    progress = (writedLen * 100 / fileSize).toInt()

                    val ae = ActionEvent(FileUtil::class.java, PROGRESS_ACTION_ID, progress.toString())
                    listener.actionPerformed(ae)
                }

                len = ins.read(buf)
            }

            out.close()
            ins.close()
        } catch (e: Exception) {
            logger.warn("", e)
            result = -1
        }

        return result
    }

    /**
     * 从路径中分离出文件名
     * @param filePathName
     * @return
     */
    fun getFileName(filePathName: String): String {
        var filePathName = filePathName
        if (StrUtil.isEmpty2(filePathName)) {
            return ""
        }

        filePathName = filePathName.trim { it <= ' ' }

        val pos1 = filePathName.lastIndexOf("/")
        val pos2 = filePathName.lastIndexOf("\\")
        val pos = if (pos1 > pos2) pos1 else pos2

        return if (pos >= 0) {
            filePathName.substring(pos + 1)
        } else {
            filePathName
        }
    }

    /**
     * 从路径中分离出路径和文件名
     * @param filePathName
     * @return
     */
    fun getFilePathName(filePathName: String): Array<String> {
        var filePathName = filePathName
        val data = arrayOf("", "")

        if (StrUtil.isEmpty2(filePathName)) {
            return data
        }

        filePathName = filePathName.trim { it <= ' ' }

        // 寻找最后一个分隔符
        val pos1 = filePathName.lastIndexOf("/")
        val pos2 = filePathName.lastIndexOf("\\")
        val pos = if (pos1 > pos2) pos1 else pos2

        //
        if (pos == filePathName.length - 1) {
            // 以分隔符结尾，一定是路径，不带文件名
            data[0] = filePathName
            data[1] = ""
        } else if (pos >= 0) {
            data[0] = filePathName.substring(0, pos + 1)
            data[1] = filePathName.substring(pos + 1)
        } else {
            data[0] = ""
            data[1] = filePathName
        }

        return data
    }

    /**
     * 取得文件扩展名
     * @param pathFileName
     * @return
     */
    fun getFileExt(pathFileName: String): String {
        val fileName = getFileName(pathFileName)

        if (StrUtil.isEmpty2(fileName)) {
            return ""
        }

        val pos = pathFileName.lastIndexOf(".")

        return if (pos >= 0) {
            pathFileName.substring(pos + 1)
        } else {
            ""
        }
    }

    /**
     * 删除多个文件
     * @param files
     */
    //	public static void deleteFiles(String files) {
    //
    //	}

    /**
     * 在路径后拼接文件名
     * @param path
     * @param fileName
     * @return
     */
    fun appendFileName(path: String, fileName: String): String {
        var path = path
        var fileName = fileName
        path = StrUtil.safeTrim2(path)
        fileName = StrUtil.safeTrim2(fileName)

        var b1 = false
        var b2 = false

        if (StrUtil.isEmpty(path)) {
            return fileName
        }

        if (StrUtil.isEmpty(fileName)) {
            return path
        }

        val suffix = path.substring(path.length - 1)
        if (suffix == "/" || suffix == "\\") {
            b1 = true
        }

        val prefix = fileName.substring(0, 1)
        if (prefix == "/" || prefix == "\\") {
            b2 = true
        }

        return if (b1 && b2) {
            path + fileName.substring(1)
        } else if (b1 || b2) {
            path + fileName
        } else {
            "$path/$fileName"
        }
    }

    /**
     * 为文件名添加扩展名
     * @param fileName
     * @param ext
     * @return
     */
    fun appendFileExt(fileName: String, ext: String): String {
        var fileName = fileName
        var ext = ext
        fileName = StrUtil.safeTrim2(fileName)
        ext = StrUtil.safeTrim2(ext)

        return if (StrUtil.isEmpty2(ext)) {
            fileName
        } else "$fileName.$ext"

    }

    /**
     * 删除单个文件
     * @param filePathName
     * @return
     */
    fun deleteFile(filePathName: String): Boolean {
        val file = File(filePathName)

        if (file.exists()) {
            if (!file.isDirectory) {
                file.delete()
                return true
            }
        }

        return false
    }

    /**
     * 读取文本文件
     * @param filePathName
     * @return
     * @throws FileNotFoundException
     * @throws IOException
     */
    @Throws(FileNotFoundException::class, IOException::class)
    fun readTextFile(filePathName: String): List<String> {
        val lines = ArrayList<String>()
        var reader: InputStreamReader? = null
        var br: BufferedReader? = null
        var input: InputStream? = null

        try {
            input = FileInputStream(filePathName)
            reader = InputStreamReader(input)
            br = BufferedReader(reader)
            var line: String? = br.readLine()

            while (line != null) {
                lines.add(line)

                line = br.readLine()
            }
        } finally {
            try {
                if (br != null) {
                    br.close()
                }
            } catch (e: IOException) {
            }

            try {
                if (reader != null) {
                    reader.close()
                }
            } catch (e: IOException) {
            }

            try {
                if (input != null) {
                    input.close()
                }
            } catch (e: IOException) {
            }

        }

        return lines
    }
}
